core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java

```diff
@@ -929,32 +938,51 @@ public PersistenceState getPersistenceState(long fileId) throws FileDoesNotExist
       inode = inodePath.getInode();
       auditContext.setSrcInode(inode);
 
-      List<FileInfo> ret = new ArrayList<>();
-      if (inode.isDirectory()) {
-        TempInodePathForDescendant tempInodePath = new TempInodePathForDescendant(inodePath);
-        try {
-          mPermissionChecker.checkPermission(Mode.Bits.EXECUTE, inodePath);
-        } catch (AccessControlException e) {
-          auditContext.setAllowed(false);
+      List<FileInfo> ret = listStatusInternal(inodePath, auditContext,
+          listStatusOptions.isRecursive(), inodePath);
+
+      auditContext.setSucceeded(true);
+      Metrics.FILE_INFOS_GOT.inc();
+      return ret;
+    }
+  }
+
+  private List<FileInfo> listStatusInternal(LockedInodePath currInodePath,
+                                                  AuditContext auditContext,
+                                                  boolean recursive,
+                                                  LockedInodePath originalInodePath)
+      throws FileDoesNotExistException, UnavailableException, AccessControlException {
+    List<FileInfo> statusList = new ArrayList<>();
+    Inode<?> inode = currInodePath.getInode();
+    if (inode.isDirectory()) {
+      try {
+        mPermissionChecker.checkPermission(Mode.Bits.EXECUTE, currInodePath);
+      } catch (AccessControlException e) {
+        auditContext.setAllowed(false);
+        if (recursive) {
+          return statusList;
+        } else {
           throw e;
         }
+      }
+      if (recursive || currInodePath.equals(originalInodePath)) {
         for (Inode<?> child : ((InodeDirectory) inode).getChildren()) {
-          child.lockReadAndCheckParent(inode);
-          try {
-            // the path to child for getPath should already be locked.
-            tempInodePath.setDescendant(child, mInodeTree.getPath(child));
-            ret.add(getFileInfoInternal(tempInodePath));
-          } finally {
-            child.unlockRead();
+          try (LockedInodePath childInodePath = mInodeTree.lockFullInodePath(child.getId(),
+              InodeTree.LockMode.READ)) {
+            statusList.addAll(listStatusInternal(childInodePath, auditContext,
+                recursive, originalInodePath));
+          } catch (Exception e) {
+            LOG.debug("ListStatus failed for path {}", mInodeTree.getPath(child).getPath(), e);
+            continue;
           }
         }
-      } else {
-        ret.add(getFileInfoInternal(inodePath));
       }
-      auditContext.setSucceeded(true);
-      Metrics.FILE_INFOS_GOT.inc();
-      return ret;
     }
+    // We add the file info of all files and any directory that is not the original directory
+    if (currInodePath.getInode().isFile() || !currInodePath.equals(originalInodePath)) {
```

Comment:
why are only files added?

Also, can we just add all the paths to the list? If we do a preorder traversal, we would always add the input path first. If we don't want the first input path, we can detect that with an empty list.

Reply:
Files and directories are added. The only exception is when the original inode is a directory, then we don't add it to the list.  I am not sure how we can throw away the first input path with an empty list?

Reply:
Well, if we always insert the input inode, and then its children, when the output list is empty, we know the first inode we can insert is the very first input inode. Therefore, we can see if we want to insert that one or not, without having to pass in `originalInodePath`.